### ISSUE TITLE
[阻塞] 请提供 VOC-20 三模型 checkpoint 与训练记录

### ISSUE THREAD

**Issue by hycx233** [oioioi-521/GO3-XAI01-04#4]
@YaoYang-123456

## 背景

成员 A 负责的 IG / Grad-CAM 六个 ImageNet 正式单元已经完成并验收；VOC 的六个对应单元目前被模型权重阻塞。台式机上的冻结数据、代码、YAML、CUDA 环境和测试均已就绪，但 `models/checkpoints/` 为空。

不能直接使用 ImageNet-1k 分类头运行 VOC：当前 VOC target 是固定顺序的 20 类索引，直接套用 1,000 类输出会让预测、归因目标和 MoRF 指标失去语义。

## 请求交付

请提供以下三个 VOC-20 checkpoint：

- `models/checkpoints/vgg16_voc20.pt`
- `models/checkpoints/resnet50_voc20.pt`
- `models/checkpoints/densenet121_voc20.pt`

文件应为完整 `state_dict`，或包装为 `{"state_dict": ...}`；须能被当前 `models/factory.py` 以 `strict=True` 加载。

分类头输出必须为 20 维，类别顺序严格为：

```text
aeroplane, bicycle, bird, boat, bottle, bus, car, cat, chair, cow,
diningtable, dog, horse, motorbike, person, pottedplant, sheep, sofa,
train, tvmonitor
```

## 训练与溯源要求

请同时提交或说明：

1. 训练代码、配置和对应 commit SHA；
2. 训练数据及切分方式；冻结的 500 张 VOC debug/eval 图片来源于 VOC2007 trainval，训练时必须全部排除，避免评估泄漏；
3. 初始化权重、损失函数、预处理、随机种子、epoch/优化器等关键设置；
4. 验证集指标和三份 checkpoint 的 SHA256；
5. 当前实验管线采用 20 类单标签 softmax。如果现有权重采用多标签 BCE/sigmoid，请先在本 Issue 说明，不能直接接入现有指标口径。

## 验收条件

- 三个 checkpoint 均可 `strict=True` 加载；
- 单张前向输出形状为 `(1, 20)`，且输出、预测和归因均为有限值；
- 类别顺序、数据排除清单和 checkpoint 哈希可追溯；
- 首图检查通过后，再按 40 张 debug → 460 张正式 eval 的顺序运行。

如果这三套权重尚未训练或无法在当前周期交付，请直接回复当前状态和预计时间；也可以在这里确认由成员 A 接管训练管线，避免继续隐式等待。

**Comment by YaoYang-123456** [comment]
已收到。目前 B 侧的三模型统一接口、预测表和 Occlusion 接入已经完成；三份 VOC-20 checkpoint 尚未训练。

我会先负责补齐训练协议和可复现链路，不会直接使用 ImageNet-1k 分类头生成 VOC 结果，也不会在标签口径未确认时直接开训。接下来按以下顺序推进：

1. 审计 VOC2007 trainval 标注与冻结清单，确保 500 张 debug/eval 图片全部从训练及验证数据中排除；
2. 根据实际数据分布比较单标签 CrossEntropy 与多标签 BCE 两种方案，优先选择与当前 20 类单标签 softmax 管线兼容、且标签语义清晰的方案；
3. 编写统一的参数化训练脚本和配置，分别训练 VGG16、ResNet50、DenseNet121；
4. 提交训练配置、数据划分、随机种子、验证指标、类别顺序、checkpoint SHA256 及严格加载验证记录；
5. checkpoint 验收后，再按首图检查 → 40 张 debug → 460 张正式 eval 的顺序运行。

我会先完成数据与训练协议审计，并在本 Issue 更新推荐方案、训练机器安排和预计交付时间。如果审计后发现当前周期或算力无法支持三模型训练，我会及时在这里说明并协调是否由 A 接管，不会继续隐式等待。

**Comment by YaoYang-123456** [comment]
请暂时保留并备份 IG/Grad-CAM 的热力图、逐图 CSV、预测表和配置快照，B 正在复核 MoRF 口径，后续可能利用现有热力图重新评分。

**Comment by YaoYang-123456** [comment]
第二阶段数据审计和 GPU 环境验证已经完成。

完整 VOC2007 trainval 共 5,011 张，冻结的 500 张 debug/eval 图片全部位于其中并已验证可以严格排除，排除后剩余 4,511 张训练候选，训练、验证与冻结评估集可以做到完全无交集。

审计发现，若采用“仅保留单类别图片 + CrossEntropy”的方案，最终只能保留 2,308 张图片，其中 `diningtable` 仅有 4 张，按训练/验证划分后训练集只有 3 张，无法形成可靠的分类器。相比之下，使用完整 4,511 张图片进行多标签 BCE 训练时，最少类别仍有 88 个正样本，数据利用率和类别覆盖明显更合理。

因此 B 建议采用 VOC 原生的多标签训练协议，并在接入现有实验管线前完成以下兼容改造：

* checkpoint 仍输出固定顺序的 20 维 logits，并支持 `strict=True` 加载；
* VOC 训练使用 `BCEWithLogitsLoss`，推理置信度使用 sigmoid；
* 冻结的 500 张评估图片本身为单标签，因此归因 target 仍使用其唯一真值类别；
* VOC 的 top-1 预测由 20 个 sigmoid/logit 分数取 argmax 得到；
* MoRF 使用有界的 raw sigmoid true-target deletion AUC，不再使用可能因低分母爆炸的概率比值；
* ImageNet 的 softmax 行为保持不变；
* 训练记录同时报告 mAP、macro/micro-F1，并在冻结单标签集上报告 top-1 accuracy。

独立 CUDA 环境已经验证完成：RTX 5080 Laptop GPU、CUDA 12.8、`torch 2.11.0+cu128`，GPU 矩阵计算、AMP 前向/反向、28 项项目测试和 `pip check` 均通过。

如无异议，我将按以上协议实现训练 manifest、sigmoid 预测与 MoRF 兼容层，然后训练 VGG16、ResNet50、DenseNet121 三份 checkpoint。预计仍在 1～2 天内完成，并提交数据排除清单、配置、验证指标和 SHA-256。

**Comment by YaoYang-123456** [comment]
VOC-20 训练协议确认请求

已完成完整 VOC2007 trainval 审计。严格排除冻结评估集 500 张后，剩余 4511 张：

单标签 CrossEntropy 协议仅保留 2308 张，其中 diningtable 仅 4 张；
多标签 BCE 协议可保留全部 4511 张，最少类别也有 88 个正样本。

因此建议将 VOC-20 checkpoint 的训练协议调整为：

20 维 multi-hot 标签；
BCEWithLogitsLoss；
推理阶段使用 sigmoid；
分类指标使用 mAP/F1；
归因仍针对指定类别；
MoRF 主指标使用有界的 raw true-target deletion AUC。

当前已在本地独立分支完成实现与测试，36 passed；仅运行了 ResNet50 的 1 epoch GPU smoke，尚未正式训练、push 或创建 PR。

请 A/C 确认是否统一采用上述多标签实验口径。确认后再进行三模型正式训练与 checkpoint 交付。

**Comment by hycx233** [comment]
A 侧确认采用 VOC 原生多标签训练协议，感谢完成数据审计。请按以下口径继续实现和正式训练：

1. 使用排除冻结 500 张图片后的 4,511 张 VOC2007 trainval 候选，20 维 multi-hot 标签与 `BCEWithLogitsLoss`；
2. VOC 推理使用 sigmoid，冻结单标签评估集的 top-1 取 20 维分数 argmax；归因 target 仍为唯一真值类别；
3. 训练报告 mAP、macro/micro-F1，并在冻结单标签集报告 top-1 accuracy；
4. checkpoint 保持 20 维固定类别顺序、`strict=True` 可加载，并交付配置、排除清单、随机种子、验证指标和 SHA-256。

MoRF 口径请同时按下面方式统一，避免 ImageNet 与 VOC 使用不同量纲：

- 新增明确命名的主指标 `faithfulness_morf_auc_raw`；
- ImageNet 使用 raw softmax true-target deletion AUC；
- VOC 使用 raw sigmoid true-target deletion AUC；
- 两者都保持在 `[0,1]`，删除比例网格保持一致；
- 现有归一化概率比值指标保留为 legacy，不要用同一 metric 名静默改变语义，也不要覆盖已完成结果；
- 预测/配置快照中记录输出激活类型（softmax 或 sigmoid），防止结果混用。

A 侧的 ImageNet `results/` 已完整备份并校验，可在兼容层合入后复用现有 2,760 张热力图重新计算 raw MoRF，无需重跑 IG/Grad-CAM 归因。备份聚合 SHA-256 为 `f1c915531f933ec47fd3a10daa451ccb98764a8111c2b348fb7de0b845652e10`。

请将训练 manifest、sigmoid 预测/MoRF 兼容层及测试 push 后创建 PR，并在此 Issue 关联。@oioioi-521 请同步确认 C 侧 RISE 管线采用相同输出激活和 `faithfulness_morf_auc_raw` 口径。

**Comment by YaoYang-123456** [comment]
成员 B 更新：VOC-20 已采用 multi-hot + BCEWithLogitsLoss + sigmoid 方案；该方向与 A 的 PR #6 交接说明及 E 的复核意见一致。\n\n新堆叠 PR：#7 https://github.com/oioioi-521/GO3-XAI01-04/pull/7 （base: feat/b-model-factory）。\n\n已实现冻结 500 张 ID 的训练/验证排除、多标签 manifest、显式 softmax/sigmoid 预测语义，以及 canonical faithfulness_morf_auc_raw；legacy faithfulness_morf_auc 仅历史保留。完整测试 37 passed、pip check 通过。\n\nResNet50 已完成 1 epoch GPU smoke，仅验证 AMP/保存/验证链路，不作为正式 checkpoint 或正式指标。后续将串行正式训练三模型；Issue #4 保持开放。

**Comment by YaoYang-123456** [comment]
正式 VOC20 多标签 checkpoint 已完成并完成本分支验收；不关闭 Issue #4，等待 A 的“单图 → 40 张 debug → 460 张 eval”验收。

- PR #7: https://github.com/oioioi-521/GO3-XAI01-04/pull/7
- 训练代码提交：`f791d1e8a346b66d21c807caabeb1a293cce84b3`
- 协议：20 维 multi-hot + `BCEWithLogitsLoss` + sigmoid；冻结 500 张 VOC 图片在 manifest 构建前排除，未参与训练、验证、早停或阈值选择。

| 模型 | 最佳 epoch | mAP | macro-F1 | micro-F1 | SHA-256 |
| --- | ---: | ---: | ---: | ---: | --- |
| ResNet50 | 30 | 0.7727 | 0.6033 | 0.6185 | `1e54867cf18ed02383ba329c944a9143871fc37fe268bec6e769a8d364ae134f` |
| DenseNet121 | 29 | 0.7883 | 0.6205 | 0.6373 | `ab61772eaf3459842f4b1e34be9ec15062e02a940d521390d80df0371168feca` |
| VGG16 | 24 | 0.8182 | 0.7378 | 0.7551 | `61640717641bfc1e9ae5d6172de667ce45755e6f046f458c848f70cb4a6c9d58` |

正式相对路径、字节大小、配置/manifest/冻结排除清单哈希见 `docs/VOC20_CHECKPOINT_MANIFEST.json`；二进制在 Git 忽略的 `models/checkpoints/`，未上传至 Git、Release 或第三方网盘。三模型均 strict-load，并在冻结 debug 图像上产生有限 `(1,20)` logits/sigmoid。

限制：ResNet50 历史 `elapsed_seconds=202.94` 仅代表最后一次恢复进程，不能作为总训练时长；新累计时长逻辑只适用于未来训练。一次 CUDA unknown error / Windows TDR 后已从上一个完整原子 checkpoint 成功恢复，根因尚未确定。若 A 无法访问本机 checkpoint，当前阻塞是缺少已批准的二进制传输渠道。

**Comment by YaoYang-123456** [comment]
B 侧复核：三份 VOC20 checkpoint 已被 `integrate/a-voc-eval` 的正式实验使用，`docs/A_IG_GRADCAM_RESULTS.md` 记录的三份 SHA-256 与 B 的 manifest 完全一致；IG/Grad-CAM 的六个 VOC 单元均已完成每单元 460 张正式运行，输出为有限值且日志为 `processed=460, skipped=0`。

因此，Issue #4 所跟踪的 checkpoint 交付与 VOC 验收阻塞在技术上已经解除。请 A 确认验收记录无遗漏后关闭本 Issue。PR #7 的代码审查与既定分支合并顺序继续单独处理，不以关闭本 Issue 代替 PR 审查。

**Comment by hycx233** [comment]
A 侧确认 checkpoint 交付与 VOC 正式验收均已完成，本 Issue 的阻塞条件已经解除。

验收证据：

- 三份 checkpoint 的 SHA-256 与 `docs/VOC20_CHECKPOINT_MANIFEST.json` 完全一致：
  - VGG16 `61640717641bfc1e9ae5d6172de667ce45755e6f046f458c848f70cb4a6c9d58`
  - ResNet50 `1e54867cf18ed02383ba329c944a9143871fc37fe268bec6e769a8d364ae134f`
  - DenseNet121 `ab61772eaf3459842f4b1e34be9ec15062e02a940d521390d80df0371168feca`
- 均已通过 `strict=True` 加载、有限 `(1,20)` 前向和 sigmoid 输出检查；
- IG/Grad-CAM 的 VOC 单图、40 张 debug、460 张 eval 均已完成；
- 六个 VOC 正式单元均为 `processed=460, skipped=0, warmup_runs=1`；
- 正式输出使用 `faithfulness_morf_auc_raw`、21 点删除网格、float32 NPY 和预热后计时；
- 完整结果说明见 `integrate/a-voc-eval` 分支的 `docs/A_IG_GRADCAM_RESULTS.md`，对应提交 `ddaa7dc`；
- 台式机正式环境完整测试为 58 passed。

结果树仍按仓库约定由 Git 忽略，并已在实验机与本机之间做聚合 SHA-256 一致性校验。PR #7 的代码审查与堆叠合并顺序继续独立处理，不以关闭本 Issue 代替 PR 审查。

因此按 completed 关闭。感谢 @YaoYang-123456 完成训练、恢复链路、manifest 与 checkpoint 交付。
